packages/contracts-periphery/contracts/universal/op-nft/Optimist.sol

```diff
@@ -0,0 +1,187 @@
+    constructor(
+        string memory _name,
+        string memory _symbol,
+        address _admin,
+        address _attestationStation
+    ) Semver(0, 0, 1) {
+        attestationStation = AttestationStation(_attestationStation);
+        initialize(_name, _symbol, _admin, _attestationStation);
+    }
+
+    /**
+     * @notice  Initialize the Optimist contract.
+     * @dev     Initializes the Optimist contract with the given parameters.
+     * @param   _name  The token name.
+     * @param   _symbol  The token symbol.
+     * @param   _admin  The administrator address.
+     * @param   _attestationStation  The address of the attestation station contract.
+     */
+    function initialize(
+        string memory _name,
+        string memory _symbol,
+        address _admin,
+        address _attestationStation
+    ) public initializer {
+        __ERC721_init(_name, _symbol);
+        __ERC721Burnable_init();
+        __Ownable_init();
+        transferOwnership(_admin);
+    }
+
+    /**
+     * @notice  Mint the Optimist token.
+     * @dev     Mints the Optimist token to the give recipient address.
+     *          Limits the number of tokens that can be minted to one per address.
+     *          The tokenId is the uint256 of the recipient address.
+     * @param   _recipient  The address of the token recipient.
+     */
+    function mint(address _recipient) public {
+        require(balanceOf(_recipient) == 0, "ALREADY_MINTED");
+        require(isWhitelisted(_recipient), "NOT_WHITELISTED");
+        _safeMint(_recipient, _tokenIdOfOwner(_recipient));
+    }
+
+    /**
+     * @notice Returns the URI for the token metadata.
+     * @dev The token URI will be stored at baseURI + '/' + tokenId + .json
+     * @param tokenId The token ID to query.
+     * @return The URI for the given token ID.
+     */
+    function tokenURI(uint256 tokenId) public view virtual override returns (string memory) {
+        require(ownerOf(tokenId) != address(0), "NOT_MINTED");
+        return
+            string(
+                abi.encodePacked(
+                    baseURI(),
+                    "/",
+                    Strings.toHexString(tokenId, _ADDRESS_LENGTH),
+                    ".json"
+                )
+            );
+    }
+
+    /**
+     * @notice  Returns whether an address is whitelisted
+     * @dev     The whitelist is an attestation by the admin of this contract
+     * @return  boolean  Whether the address is whitelisted
+     */
+    function isWhitelisted(address _recipient) public view returns (bool) {
+        return
+            attestationStation
+                .attestations(owner(), _recipient, bytes32("optimist.can-mint"))
+                .length > 0;
+    }
+
+    /**
+     * @notice  Returns decimal tokenid for a given address
+     * @return  uint256 decimal tokenId
+     */
+    function tokenIdOfOwner(address _owner) public view returns (uint256) {
+        require(balanceOf(_owner) > 0, "NOT_MINTED");
```

Comment:
Also not sure if we really need the `require` here. Should be able to get token ID even if haven't minted. Also means you can get rid of the internal function.